File: scripts/python/image_provider_ladder.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import shutil


TRUTHY = {"1", "true", "yes", "on", "y"}


def main() -> int:
    parser = argparse.ArgumentParser(description="Inspect the Yali-first image provider ladder.")
    parser.add_argument("--intent", default="generation", choices=["generation", "edit", "prompt", "ppt"])
    parser.add_argument("--execution-requested", action="store_true")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    has_python = shutil.which("python3") is not None
    has_node = shutil.which("node") is not None
    has_runtime = has_python or has_node
    has_yali_key = bool(os.environ.get("YALIAI_API_KEY", "").strip())
    allow_compat = os.environ.get("YALIAI_ALLOW_COMPAT_PROVIDER", "").strip().lower() in TRUTHY
    has_compat_key = bool(os.environ.get("OPENAI_API_KEY", "").strip())
    provider_order = ["public-yali-retrieval"]
    if args.execution_requested and has_yali_key and has_runtime:
        provider_order.append("yali-api")
    if args.execution_requested and allow_compat and has_compat_key and has_runtime:
        provider_order.append("compatible-fallback")
    if args.execution_requested:
        provider_order.extend(["host-native-fallback-if-available", "advisor"])
    else:
        provider_order.append("prompt-spec-output")
    recommended = next((item for item in provider_order if item != "public-yali-retrieval"), "public-yali-retrieval")
    result = {
        "ok": True,
        "intent": args.intent,
        "execution_requested": args.execution_requested,
        "has_python3": has_python,
        "has_node": has_node,
        "has_runtime": has_runtime,
        "public_yali_retrieval": True,
        "has_yaliai_key": has_yali_key,
        "compatible_fallback_allowed": allow_compat,
        "has_compatible_key": has_compat_key,
        "compatible_base_url": os.environ.get("OPENAI_BASE_URL") or "https://api.openai.com/v1",
        "compatible_model": os.environ.get("OPENAI_IMAGE_MODEL") or "gpt-image-1",
        "provider_order": provider_order,
        "recommended_mode": recommended,
    }
    if args.json:
        print(json.dumps(result, ensure_ascii=False, indent=2))
    else:
        print("--- Yali-first image provider ladder ---")
        for key, value in result.items():
            print(f"{key}: {' -> '.join(value) if isinstance(value, list) else value}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/node/image_provider_ladder.mjs
#!/usr/bin/env node
import { spawnSync } from "node:child_process";

const TRUTHY = new Set(["1", "true", "yes", "on", "y"]);
const USAGE = `Usage:
  node scripts/node/image_provider_ladder.mjs [--intent generation|edit|prompt|ppt] [--execution-requested] [--json]`;

function commandExists(command) {
  return spawnSync(process.platform === "win32" ? "where" : "sh", process.platform === "win32" ? [command] : ["-lc", `command -v ${command}`], { encoding: "utf8" }).status === 0;
}

function parseArgs(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    console.log(USAGE);
    process.exit(0);
  }
  const args = { intent: "generation", executionRequested: false, json: false };
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (item === "--execution-requested") {
      args.executionRequested = true;
      continue;
    }
    if (item === "--json") {
      args.json = true;
      continue;
    }
    if (item === "--intent") {
      args.intent = argv[++i] || "generation";
      continue;
    }
    throw new Error(`Unexpected argument: ${item}`);
  }
  return args;
}

const args = parseArgs(process.argv.slice(2));
const hasPython = commandExists("python3");
const hasNode = commandExists("node");
const hasRuntime = hasPython || hasNode;
const hasYaliKey = Boolean(String(process.env.YALIAI_API_KEY || "").trim());
const allowCompat = TRUTHY.has(String(process.env.YALIAI_ALLOW_COMPAT_PROVIDER || "").trim().toLowerCase());
const hasCompatKey = Boolean(String(process.env.OPENAI_API_KEY || "").trim());
const providerOrder = [
  "public-yali-retrieval",
  ...(args.executionRequested && hasYaliKey && hasRuntime ? ["yali-api"] : []),
  ...(args.executionRequested && allowCompat && hasCompatKey && hasRuntime ? ["compatible-fallback"] : []),
  ...(args.executionRequested ? ["host-native-fallback-if-available", "advisor"] : ["prompt-spec-output"]),
];
const recommendedMode = providerOrder.find((item) => item !== "public-yali-retrieval") || "public-yali-retrieval";
const result = {
  ok: true,
  intent: args.intent,
  execution_requested: args.executionRequested,
  has_python3: hasPython,
  has_node: hasNode,
  has_runtime: hasRuntime,
  public_yali_retrieval: true,
  has_yaliai_key: hasYaliKey,
  compatible_fallback_allowed: allowCompat,
  has_compatible_key: hasCompatKey,
  compatible_base_url: process.env.OPENAI_BASE_URL || "https://api.openai.com/v1",
  compatible_model: process.env.OPENAI_IMAGE_MODEL || "gpt-image-1",
  provider_order: providerOrder,
  recommended_mode: recommendedMode,
};

if (args.json) console.log(JSON.stringify(result, null, 2));
else {
  console.log("--- Yali-first image provider ladder ---");
  for (const [key, value] of Object.entries(result)) console.log(`${key}: ${Array.isArray(value) ? value.join(" -> ") : value}`);
}
